SC-XXX: Improved speed of docker image building

sprymiker opened the pull request:
### Description

- Ticket/Issue: TBC

Inspired by https://github.com/spryker/docker-sdk/tree/onsteroids. Thank you, @oleg-glushak!
Difference:
- Incorporated as a general building mechanism, not just alternative
- More efficient dependency tree to increase parallelization and decrease time-to-done
- Pipeline image contains dev dependency as it is required
- Parallel pre-pull and post-push are not necessary as `docker build bake` covers this OOTB
- External cache sources are not necessary for now as it does not bring any significant time benefits

P.S. Image size concern was not fully considered. Additional activities measurements are needed, including changes in [spryker/docker-php](https://github.com/spryker/docker-php) as a separate topic.

#### Related resources

- https://docs.docker.com/build/building/multi-stage/
- https://docs.docker.com/engine/reference/commandline/buildx_bake/

#### Change log

Improved speed of docker image building

- Introduced `docker bake` as the primary build mechanism with a fallback to `docker build` if bake is not available
- Adjusted build process by utilizing multistage builds and paralleling builds
- Optimised cache usage by separating/moving parameters that are changed often (e.g. build hash, stamp)
- Fixed bug with `docker/sdk export` when export options were ignored when `-v` option was used at the same time
- Added labels for all images containing project revision, project name, and SDK revision
- Adjusted maintenance mode to be controlled by environment variable: `SPRYKER_MAINTENANCE_MODE_ENABLED=1 docker/sdk run`
- Adjusted default progress type to auto for more compatibility

### Checklist
- [X] I agree with the Code Contribution License Agreement in CONTRIBUTING.md

sprymiker committed:
Improved speed of docker image building
- Introduced `docker bake` as main build mechanism with fallback to `docker build` if bake is not available
- Adjusted build process by utilising multistage builds and paralleling builds
- Optimised cache usage by separating/moving parameters that are changed often (e.g. build hash, stamp)
- Fixed bug with `docker/sdk export` when export options was ignored when `-v` option was used at the same time
- Added labels for all images containing project revision, project name and SDK revision
- Adjusted maintenance mode to be controlled by environment variable: SPRYKER_MAINTENANCE_MODE_ENABLED=1 docker/sdk run

sprymiker committed:
Moved dockerfiles to match current structure and keep git history as update instead of create/delete

sprymiker committed:
Adjusted default progress type to auto for more compatibility

sprymiker committed:
Replaced dependency on `ip` package in linux to use more common packages

sprymiker committed:
Fixed wrong path appeared after merging with master

oleg-glushak commented:
Great to see this being worked on @sprymiker.
I'm not entirely sure, if `External cache sources are not necessary for now as it does not bring any significant time benefits` can be correct. Have we measured how long it takes for `sdk export images -d ecr` command to complete when running in the Cloud (Codebuild) with these changes?

For comparison: `sdk export images-slim -d ecr-slim` command using `https://github.com/spryker/docker-sdk/tree/onsteroids` running in Codebuild takes ~8m30s for [spryker/suite-nonsplit](https://github.com/spryker/suite-nonsplit)

sprymiker committed:
Adjusted bake configuration by separating targets in export process allowing parallel pushes of application images

sprymiker committed:
- Added fallback for cases when buildx docker plugin is not installed
- Removed intermediate dump-autoload as it probably takes more time than benifits for the following build

sprymiker committed:
- Reduced requirement for buildx version as it is confirmed working on 0.7.1
- Upgraded syntax docker version
- Removed parallelisation for fallback build (without buildx)

sprymiker committed:
- Fixed mount cache permissions and wrong id

sprymiker committed:
Fixed git cleanup command (mostly for nonsplit)

sprymiker committed:
Fixed git cleanup command (mostly for nonsplit). Part 2

sprymiker committed:
Deep speed, number of layers and image size optimizations. Restructuring dependencies.

sprymiker committed:
Speeding up dev composer run

sprymiker committed:
Fixed assets build process that does not build

sprymiker committed:
Excluding tests/dd.php to be skipped for web app images

sprymiker committed:
Reduced number of layers in web service images

sprymiker committed:
Fixed autoload issue with spryker/spryker

sprymiker committed:
Reduced pipeline image by excluding git folders in vendor, but keeping them in cli image

sprymiker committed:
Fixed resolver cache for cli/pipeline

sprymiker committed:
Fixed cli and scheduler does not work properly

sprymiker committed:
Fixed permission issue in cli

sprymiker committed:
Optimized pipeline build speed by reusing application build results

sprymiker committed:
Optimized pipeline build speed by reusing application build results [Part 2]

sprymiker committed:
Fixed lack of files in root folder for FE build in suite-nonsplit (angular)

sprymiker committed:
Improved docker build structure and hopefully speed be using --mount=type=bind

sprymiker committed:
Refactored Dockerfiles to use newest build features
Optimized pipeline and jenkins images by removing unnecessary binaries (e.g. PGSQL client if database is MYSQL)

oleg-glushak commented:
Any chances for this PR to be merged eventually @zyuzka @alexanderM91

alexanderM91 commented:
> Any chances for this PR to be merged eventually @zyuzka @alexanderM91

Hi @oleg-glushak, 
let's wait for an answer from the product team
